### ISSUE TITLE
Evaluating Overfitting for Offline Testing -Deep Learning Recommendation Feature Models

### ISSUE THREAD

**Issue by Shadz13** [massquantity/LibRecommender#28]
In deep learning, we split the data into a training set and validation OR test set. We evaluate overfitting by observing the accuracy and loss results from the training set and validation set over each epoch. We can also evaluate the unrepresented data with these results.

How does one evaluate overfitting for Recommendation Systems? Do we simply split the data into a train, test and validation set?  However, this will decrease our sample size and in my case I am limited with real data. Is showcasing the accuracy and loss results outputted from the validation set for each epoch sufficient? Therefore, we do not test for overfitting but ensure that our results meet the recommendation models requirements in terms of establishing the balance between popularity, specific wants and generalization etc. This will be evaluated using online A/B testing or integrating a filter module. 

In addition, I noticed that the scripts do not have an early stop. Therefore, this may be a solution how does one implement this on a high level?

**Comment by massquantity** [comment]
Indeed, accuracy and loss outputs are not sufficient for evaluating recommendation systems. But they are the most common way, and other considerations such as popularity etc are related to specific domains, so I didn't think much about it.

If you have limited data and don't care about the validation result, then just use all the data to build the trainset without splitting data first. Also don't set the `eval_data` argument in the `fit()` function.

As for early stopping, well technically, it is coupled with a validation set to _stop_ training when validation loss increases. 
I didn't add early stopping to the library because: 1) It will make the code looks ugly. 2) When I first get some data, the most probable thing I'll do is training the data multiple times to find best hyper-parameters. Then after I've found the most suitable parameters, I've pretty much known which epoch to stop, so at last I don't need early stopping anymore.
Right now I'm not clear how to implement it in detail so I can't give you advice.
